### PR TITLE
Backpropagate stack pointer from procedure return.

### PR DESCRIPTION
**Pull request by ptomin**
Assume that stack pointer at the end of procedure has the same
value as at the start

If we have
```
    call eax; We do not know calling convention of this indirect call
            ; So we do not know value of stack pointer after it
cleanup:
    pop esi
    pop ebp
    ret
```
then we could assume than stack pointer at `cleanup` label is
`fp - 8`

### CODE REVIEW

File: subjects/PE/x86/VCExeSample/VCExeSample.c [hunk 1]
```diff
@@ -92,36 +91,24 @@ void const_div_test10(word32 dwArg04)
...
 		else
-		{
-			*(fp - 0x14) = ebp_19->r0008;
-			ebp_19 = nested_if_blocks_test8(rArg00, rArg0);
-		}
-		word32 ecx_38 = *(ebp_19 - 0x04);
-		*(ebp_19 - 0x04) = ecx_38 - 0x01;
-		ecx = ecx_38 - 0x01;
+			nested_if_blocks_test8(rArg04, rArg0);
+		dwLoc08_28 = dwLoc08_28 - 0x01;
```
Comment (ptomin) [inline subjects/PE/x86/VCExeSample/VCExeSample.c]: This is big improvement. Is not it, John?

File: subjects/PE/MIPS/swlswr/test.c [hunk 2]
```diff
@@ -232,10 +232,5 @@ code * fn0001152C(code * r2, word32 dwArg00)
 // 000116FC: void fn000116FC(Register (ptr code) r2, Stack word32 dwArg00)
 void fn000116FC(code * r2, word32 dwArg00)
 {
-	if (fn0001152C(r2, dwLoc20) != 0x00)
-		fp->dwFFFFFFF4 = 0x00;
-	else
-		fp->dwFFFFFFF4 = -0x01;
-	fp->dwFFFFFFF0 = fp->dwFFFFFFF4;
```
Comment (ptomin) [inline subjects/PE/MIPS/swlswr/test.c]: This is regression. But it' another issue which should not prevent to merge this PR.
`ControlFlowGraphCleaner` eliminates `fn0001152C` call. Possible fix is 2d6f8ee

**Comment by ptomin** [comment]
I have created unit test (see 6875bf4). Is it enought, John?

File: subjects/PE/x86/VCExeSample/VCExeSample.c (hunk 1 above)
Comment (uxmal) [reply subjects/PE/x86/VCExeSample/VCExeSample.c]: Absolutely. We don't want to lose calls after all!

File: src/Decompiler/Analysis/TrashedRegisterFinder.cs [hunk 3]
```diff
@@ -459,6 +463,126 @@ private bool ProcedureTerminates(Expression expr)
...
+            var tes = new TrashedExpressionSimplifier(
+                this.program.SegmentMap, this, ctx);
+            this.se = new SymbolicEvaluator(tes, ctx);
+        }
+
+        private bool GetOffset(Expression e, Identifier id, out int offset)
+        {
+            offset = 0;
+            if (e is BinaryExpression bin)
+            {
```
Comment (uxmal) [inline src/Decompiler/Analysis/TrashedRegisterFinder.cs]: I would check for `Operator.ISub` and `Operator.IAdd` just to make sure.

File: src/Decompiler/Analysis/TrashedRegisterFinder.cs [hunk 4]
```diff
@@ -459,6 +463,126 @@ private bool ProcedureTerminates(Expression expr)
...
+
+        private void SetValue(
+            SymbolicEvaluationContext ctx,
+            Storage stg,
+            Expression value)
+        {
+            ctx.RegisterState[stg] = value;
+        }
+
+        private void ClearEvaluationState(Block block)
```
Comment (uxmal) [inline src/Decompiler/Analysis/TrashedRegisterFinder.cs]: The naming had me a little confused here. Consider using `CreateEvaluationState` instead to clarify the intent.

File: src/Decompiler/Analysis/TrashedRegisterFinder.cs [hunk 5]
```diff
@@ -459,6 +463,126 @@ private bool ProcedureTerminates(Expression expr)
...
+                    offset = -offset;
+                return true;
+            }
+            else
+            {
+                return e == id;
+            }
+        }
+
+        private void BackpropagateStackPointer(Block block)
```
Comment (uxmal) [inline src/Decompiler/Analysis/TrashedRegisterFinder.cs]: Consider commenting this procedure with a description of your algorithm.

File: src/Decompiler/Analysis/TrashedRegisterFinder.cs (hunk 4 above)
Comment (ptomin) [reply src/Decompiler/Analysis/TrashedRegisterFinder.cs]: Done (rebased)

File: src/Decompiler/Analysis/TrashedRegisterFinder.cs (hunk 3 above)
Comment (ptomin) [reply src/Decompiler/Analysis/TrashedRegisterFinder.cs]: I have added this check (see rebased version)

File: src/Decompiler/Analysis/TrashedRegisterFinder.cs [hunk 6]
```diff
@@ -459,6 +463,129 @@ private bool ProcedureTerminates(Expression expr)
...
+        private bool GetOffset(Expression e, Identifier id, out int offset)
+        {
+            offset = 0;
+            if (e is BinaryExpression bin)
+            {
+                if (bin.Left != id)
+                    return false;
+                var op = bin.Operator;
+                if (op != Operator.ISub && op != Operator.IAdd)
+                    return false;
```
Comment (ptomin) [inline src/Decompiler/Analysis/TrashedRegisterFinder.cs]: This is a requested check

File: src/Decompiler/Analysis/TrashedRegisterFinder.cs [hunk 7]
```diff
@@ -459,6 +463,146 @@ private bool ProcedureTerminates(Expression expr)
...
+        ///     - if stack pointer is trashed (usually after indirect calls)
+        ///     then pass fake 'currSp' identifier as stack pointer value to
+        ///     evaluation context and remember current
+        ///     - evaluate statement
+        /// At the end of block read stack pointer value from evaluation
+        /// context. It should be like 'currSp + offset'. We assume that stack
+        /// pointer at the end is 'fp'. So 'curSp' is 'fp - offset'. So we
+        /// assume that stack pointer at last remembered statement is
+        /// `fp - offset` and store this value to block flow
+        /// </remarks>
```
Comment (ptomin) [inline src/Decompiler/Analysis/TrashedRegisterFinder.cs]: This is requested description of my algorithm. Is it enough?

File: src/Decompiler/Analysis/TrashedRegisterFinder.cs (hunk 5 above)
Comment (ptomin) [reply src/Decompiler/Analysis/TrashedRegisterFinder.cs]: See https://github.com/uxmal/reko/pull/633/files#r187793329

**Comment by ptomin** [comment]
I have done requested changes, Anything else, John?

**Comment by uxmal** [comment]
I'm going to pull this now, but I'm also going to ask you to look into merging this into the `analysis-development` branch. The trashed register analysis in that branch looks very different from the `master` branch.

**Comment by ptomin** [comment]
> but I'm also going to ask you to look into merging this into the analysis-development branch. The trashed register analysis in that branch looks very different from the master branch.

@uxmal Finally I have started implementation  of stack backpropagation for `analysis-development`. See https://github.com/ptomin/reko/tree/backpropagate-stack-pointer branch. I'll create PR when I'll finish this work

**Comment by uxmal** [comment]
I will look at this tonight my time. I'm feeling some excitement because it seems `analysis-development` is starting to stabilize.

**Pull request merged**